homeassistant/components/sensor/fail2ban.py

```diff
@@ -0,0 +1,144 @@
+    for jail in jails:
+        device_list.append(BanSensor(name, jail, log_parser))
+
+    async_add_devices(device_list, True)
+
+
+class BanSensor(Entity):
+    """Implementation of a fail2ban sensor."""
+
+    def __init__(self, name, jail, log_parser):
+        """Initialize the sensor."""
+        self._name = '{} {}'.format(name, jail)
+        self.jail = jail
+        self.ban_dict = {STATE_CURRENT_BANS: [], STATE_ALL_BANS: []}
+        self.last_ban = None
+        self.log_parser = log_parser
+        _LOGGER.debug("Setting up jail %s", self.jail)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of the fail2ban sensor."""
+        return self.ban_dict
+
+    @property
+    def state(self):
+        """Return the most recently banned IP Address."""
+        return self.last_ban
+
+    def update(self):
+        """Update the list of banned ips."""
+        if self.log_parser.timer():
+            self.log_parser.read_log(self.jail)
+
+        if self.log_parser.data:
+            for entry in self.log_parser.data:
+                _LOGGER.debug(entry)
+                current_ip = entry[1]
+                if entry[0] == 'Ban':
+                    if current_ip not in self.ban_dict[STATE_CURRENT_BANS]:
+                        self.ban_dict[STATE_CURRENT_BANS].append(current_ip)
+                    if current_ip not in self.ban_dict[STATE_ALL_BANS]:
+                        self.ban_dict[STATE_ALL_BANS].append(current_ip)
+                    if len(self.ban_dict[STATE_ALL_BANS]) > 10:
+                        self.ban_dict[STATE_ALL_BANS].pop(0)
+
+                elif entry[0] == 'Unban':
+                    if current_ip in self.ban_dict[STATE_CURRENT_BANS]:
+                        self.ban_dict[STATE_CURRENT_BANS].remove(current_ip)
+
+        if self.ban_dict[STATE_CURRENT_BANS]:
+            self.last_ban = self.ban_dict[STATE_CURRENT_BANS][-1]
+        else:
+            self.last_ban = 'None'
+
+
+class BanLogParser(object):
+    """Class to parse fail2ban logs."""
+
+    def __init__(self, interval, log_file):
+        """Initialize the parser."""
+        self.interval = interval
+        self.log_file = log_file
+        self.data = list()
+        self.last_update = dt_util.now()
+
+    def timer(self):
+        """Check if we are allowed to update."""
+        boundary = dt_util.now() - self.interval
+        if boundary > self.last_update:
+            self.last_update = dt_util.now()
+            return True
+        return False
+
+    def read_log(self, jail):
+        """Read the fail2ban log and find entries for jail."""
+        self.data = list()
+        try:
+            with open(self.log_file, 'r', encoding='utf-8') as file_data:
+                self.data = re.findall(
+                    r'\[' + re.escape(jail) + r'\].(Ban|Unban) ([\w+\.]{3,})',
```

Comment:
Compile the regex into a dict and use the jail as key. Use also this syntax: `r"...{}...".format()`

Reply:
no, like `self.xy[jail] = re.compile(r"...{}...".format(jail))`